cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java

```diff
@@ -372,19 +372,33 @@ private void writeRow(Tuple2<DecoratedKey, Object[]> keyAndRowData,
     private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, String[] columnNames, Object[] values)
     {
         Preconditions.checkArgument(values.length == columnNames.length,
-                                    "Number of values does not match the number of columns " + values.length + ", " + columnNames.length);
+                "Number of values does not match the number of columns " + values.length + ", " + columnNames.length);
 
         for (int i = 0; i < columnNames.length; i++)
         {
-            if (cqlTable.containsUdt(columnNames[i]))
+            String columnName = columnNames[i];
+            Object columnValue = values[i];
+
+            // convert all BridgeUDTValue types in the column to UDTValue
+            if (cqlTable.containsUdt(columnName))
             {
-                map.put(columnNames[i], maybeConvertUdt(values[i]));
+                columnValue = maybeConvertUdt(columnValue);
             }
-            else
+
+            // Find CqlTuple associated with the field and convert value to TupleValue
+            // Note: Tuples do not need recursive conversions
+            //       Tuple with UDT come here as Object[...UDTValue...]
+            //       Tuple with Tuple come here as Object[...Object[]...]
```

Comment:
@yifan-c I did mention here how nested tuples are handled, tuple inside tuples and tuple inside collections do not need special conversion. Only UDTs require special conversion, so we call maybeConvertUdt above if contains UDT even before reaching here. By the time we are here, either we have object or object arrays or already converted UDTs.